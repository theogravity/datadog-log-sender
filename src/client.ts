import TLS from 'tls'

import { DataDogLogSenderConfig, IDataDogLogSender } from './interfaces'

const DEFAULT_DD_HOST = 'intake.logs.datadoghq.com'
const DEFAULT_DD_PORT = 10516

export class DataDogLogSender implements IDataDogLogSender {
  protected client: TLS.TLSSocket
  protected _config: DataDogLogSenderConfig
  protected _connectCalled: boolean

  constructor (config: DataDogLogSenderConfig) {
    if (!config.apiKey) {
      throw new Error('DataDogLogSender apiKey config not defined')
    }

    this.client = null
    this._config = config
    this._connectCalled = false
  }

  /**
   * Initializes the connection to DataDog. Must be called once before sending.
   */
  async init (): Promise<TLS.TLSSocket> {
    return new Promise(resolve => {
      this.client = TLS.connect(
        {
          host: DEFAULT_DD_HOST,
          port: DEFAULT_DD_PORT,
          rejectUnauthorized: true,
          ...this._config
        },
        () => {
          this.client.setKeepAlive(true)
          this.client.setEncoding('utf-8')
          this._connectCalled = true
          resolve(this.client)
        }
      )

      if (this._config.onError) {
        this.client.on('error', err => {
          this._config.onError(err, this.client)
        })
      }
    })
  }

  protected assembleData (logData: string) {
    return `${this._config.apiKey} ${logData}`
  }

  /**
   * Sends log data to datadog by stringify-ing the object.
   */
  sendObject<T = Record<string, any>> (logData: T): this {
    return this.sendText(JSON.stringify(logData))
  }

  /**
   * Sends log data to datadog using a text string. Do *not* include a
   * newline at the end of your text.
   */
  sendText (logData: string): this {
    if (!this._connectCalled) {
      throw new Error('Call DataDogLogSender#init() first before sending data')
    }

    this.client.write(this.assembleData(logData) + '\n')
    return this
  }

  /**
   * Ends the connection to DataDog
   */
  close () {
    this.client.end()
    this.client = null
  }
}
