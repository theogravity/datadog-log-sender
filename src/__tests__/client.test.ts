/* eslint-env jest */
import TLS from 'tls'
import selfSigned from 'openssl-self-signed-certificate'
import { DataDogLogSender } from '../client'

const TEST_KEY = 'test-key'

const PORT = 10516
const HOST = 'localhost'

const { key, cert } = selfSigned

const queue = []

let server

beforeAll(done => {
  server = TLS.createServer(
    {
      key,
      cert
    },
    socket => {
      socket.on('data', data => {
        queue.push(data.toString())
      })
    }
  )

  server.listen(PORT, HOST, function () {
    console.log('Mock socket server listening at %s, on port %s', HOST, PORT)
    done()
  })
})

afterAll(() => {
  server.close()
})

describe('Datadog log sending client', () => {
  it('should send data', async done => {
    const logSender = new DataDogLogSender({
      host: HOST,
      port: PORT,
      apiKey: TEST_KEY,
      ca: cert
    })

    const socket = await logSender.init()

    socket.on('end', () => {
      expect(queue).toMatchSnapshot()
      done()
    })

    logSender.sendText('Unstructured log data')

    logSender.sendObject({
      message: 'JSON formatted log sent through TLS',
      ddtags: 'env:dd-send-log',
      ddsource: 'terminal',
      hostname: 'gs-hostame',
      service: 'user'
    })

    logSender.sendObject({
      message: 'JSON formatted log sent through TLS 2',
      ddtags: 'env:dd-send-log',
      ddsource: 'terminal',
      hostname: 'gs-hostame',
      service: 'user'
    })

    logSender.close()
  })

  it('should throw if the apiKey is not defined', () => {
    expect(() => {
      // @ts-expect-error
      new DataDogLogSender({})
    }).toThrowError()
  })

  it('should throw if init is not called', () => {
    const c = new DataDogLogSender({ apiKey: 'abcd' })
    expect(() => {
      c.sendText('test')
    }).toThrowError()
  })

  it('should throw if the CA is invalid', async done => {
    const logSender = new DataDogLogSender({
      host: HOST,
      port: PORT,
      apiKey: TEST_KEY,
      onError: (err, socket) => {
        expect(err).toBeDefined()
        done()
      }
    })

    await logSender.init()
  })
})
