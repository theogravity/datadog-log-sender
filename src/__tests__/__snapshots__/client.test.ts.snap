// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datadog log sending client should send data 1`] = `
Array [
  "test-key Unstructured log data
",
  "test-key {\\"message\\":\\"JSON formatted log sent through TLS\\",\\"ddtags\\":\\"env:dd-send-log\\",\\"ddsource\\":\\"terminal\\",\\"hostname\\":\\"gs-hostame\\",\\"service\\":\\"user\\"}
test-key {\\"message\\":\\"JSON formatted log sent through TLS 2\\",\\"ddtags\\":\\"env:dd-send-log\\",\\"ddsource\\":\\"terminal\\",\\"hostname\\":\\"gs-hostame\\",\\"service\\":\\"user\\"}
",
]
`;
